src/ide/tgui.ts

```diff
@@ -1146,14 +1167,401 @@ export let tgui = (function() {
 						return false;   // important: consume clicks
 					},
 			});
+
+	// See `createModal` for a detailed description of the objects
+	// that are stored here
 	let modal = new Array();
-	
+
+	function roundToPhysicalPixel(virtual_px: number)
+	{
+		return (virtual_px*window.devicePixelRatio | 0)/window.devicePixelRatio;
+	}
+
+	// Center a dialog at the center of the window
+	// Dialogs are aligned at integral pixel coordinates to
+	// avoid pixel-jittering
+	function centerModalDialog(dlg)
+	{
```

Comment:
[Again](https://github.com/TGlas/tscript/pull/75#pullrequestreview-589987254), this should _definitely_ be done in CSS.
Something like this would work:

```css
.tgui-modal {
	  width: ...%;
	  height: ...%;
	  min-width: ...;
	  min-height: ...;
	  position: absolute;
	  top: 50%;
	  left: 50%;
	  tranform: translate(-50%, -50%);
}
```

However, there would be some issues when the screen is too small for the `min-`size to fit. To get this 100% right with CSS, we'd need to change the DOM structure. I think this is out of scope for this PR, so I don't want to block merging with this.

Reply:
I don't think that a programming IDE needs to work on a device without a proper keyboard. Sure, that's not well defined. Anyway, assuming a minimal screen size of, say, 1280x800 (12 years old Macbook) seems safe to me. If you object, about 20 years back it was 1024x768. That's good enough for a dialog. No need to go down to 320x200, that was 28 years ([DooM](https://en.wikipedia.org/wiki/Doom_(1993_video_game))), and even 39 years ([C64](https://en.wikipedia.org/wiki/Commodore_64)) ago. You may be out of luck on a phone. Don't code on a phone :)

More seriously: If you have a good way of solving this then I'd appreciate, but I don't think that we should bother too much.

Reply:
That's exactly why I think this little CSS would be enough. Doing it in JS just to get it subpixel-perfect seems pretty overkill to me 😄

Reply:
I think if I wouldn't have worked on the CSS of the dialogs, I wouldn't have noticed pixel jittering, which appears when the browser size changes. I think no one would have ever noticed it :D. It would definitely be much cleaner code if the pixel perfect alignment was only done in CSS, but in CSS there is no possible way to do this alignment. In Firefox the pixel jittering is very noticeable, in Chrome and probably other browsers, it is not as visible. I also think it is a little overkill now. So should I discard this code and replace it back with CSS, similar to as it is in the main branch now? What do you think?

Reply:
Do you mean by "pixel jittering" that the dialog jumps around a little _while_ resizing the browser window? I think this is unfortunately normal behaviour and can't really be "fixed". For me it's even more noticeable in the JS version, since it doesn't run faster than the browser doing CSS layout.

If you mean that it's not positioned on pixel boundaries and causes anti-aliasing (does it even?), then you shouldn't worry about that at all, since it's highly dependent on the browser's rendering infrastructure. Trying to "fix" that will probably cause more problems in the long term than it will solve.

For what to do in CSS, you could either use my proposal from above, or stick with how it's currently done. I think both solutions should work fine...